Improvements on changelog to be able to deliver reports

jmelis opened the pull request:
This PR aims to fix several issues that were present:

* Refactor to separate rendering from rest of code
* Services are grouped by url and they're only checked out once
* Add multiple output formats
* Fix problem with services that have 'none' as hash
* Improve efficiency by eliminating redundant actions
* General readability issues
* Fetch repository url to be displayed in report

jmelis committed:
Changelog diff method reports only changed hashes.

Also the legibility has been improved.

jmelis committed:
Old and new commit for changelog now can be a date

jmelis committed:
Remove unnecessary argument from checkout method

jmelis committed:
Improve markdown rendering:

- Changed services are deduped by url
- Changelog header has been removed as it was wrong

jmelis committed:
Display commit change for the saas-* repo

jmelis committed:
Skip when the hash of the service is not defined

jmelis committed:
Improve markdown rendering and code readability

jmelis committed:
Refactor the code so rendering is separated from changelog

jmelis committed:
Refactor markdown to base it off of templates

jmelis committed:
Add plain formatter

jmelis committed:
Fine tune format to send in email

jaseemabid commented on saasherder/changelog.py:
```diff
@@ -1,140 +1,186 @@
+                        .format(cmd, p.returncode))
 
-class Changelog(object):
+    return stdout.decode('utf-8')
```
+1. This looks neater than the static method.

jmelis replied:
I was thinking of putting it in a separate module and doing a few more convenience methods. For instance, with this method we're are losing stderr...

jaseemabid commented on saasherder/changelog.py:
```diff
@@ -1,140 +1,186 @@
-    @staticmethod
-    def diff(now, previous):
-        """Return the difference between 2 service lists
+        section_template = textwrap.dedent("""\
```
TIL textwrap. +1

jmelis replied:
I also TIL'd just a couple of days ago! it's great and it's standard library!

jaseemabid commented on saasherder/changelog.py:
```diff
@@ -1,140 +1,186 @@
-        cmd = "cd {} && git log -1 --format=%cd {}".format(worktree, commit)
+        # Key: url of the service
+        # Value: list of services that have changed
+        self.changed_services = []
```
You are using an array and a hash with the same name. Could you simplify this a bit maybe?

jmelis replied:
Agreed.

jaseemabid commented on saasherder/changelog.py:
```diff
@@ -1,140 +1,186 @@
+
+                    changed_services[url] = diff_item
+
+        self.changed_services = changed_services.values()
```
I'd avoid this line if I can - mutating variables in a different scope.

jmelis replied:
I'm unsure what you mean? Can you further explain?

jaseemabid commented on saasherder/changelog.py:
```diff
@@ -148,40 +194,54 @@ def generate(self, context, old, new):
+            dt_old = parse(old)
+            old = run("git rev-list -1 --before='{}' {}".format(dt_old, branch)).strip()
+        except:
+            pass
```
1. Except a more specific class of Error here.
2. Should you `exit(1)` instead of pass?

jmelis replied:
I believe this is correct, `old` and `new` can be one of three things:

- a branch, for example: master
- a commit (short or long), for example: c589a14
- a date, for example: 2018-02-21 10:00

This piece of code will try to detect if it's a date (by doing dateutils.parse). If that parsing fails it shouldn't do anything and keep the original value as it's a commit or a branch. However if the parsing succeeds, then we need to find the commit associated to that time.

However I will refactor this a bit and add some comments, if it wasn't clear when reading the code it means that I need to fix it :)

jmelis replied:
You are right about excepting the specific class of Error, I hadn't read that. Will do.

jaseemabid reviewed (changes requested):
Requesting minor changes, but the patch looks quite good! I like the small functions :+1: but would avoid the mutations if I could.

jmelis committed:
Address PR comments:

- Refactor date parsing for commit
- Change variables to avoid confusion
- Use specific errors instead of except'ing the world

jaseemabid reviewed (approved):
LGTM :100: :+1:
